### PR TITLE
Add a note about aborts encountered while sending early_data

### LINKED ISSUE #6735: Document application level retry when 0-RTT connection failed to negotiate TLS 1.3
if in the connection client advertised version of protocol that is not supported by server (at 1a50eedf2a1fbb1e that means 0x0304 (TLS1.3 final) or 0x7f1d (127, 29; draft-29)) but includes 0-RTT, the connection will negotiate TLS 1.2 and server will fail to handle the early data correctly (i.e. it won't continue the connection)

see also #6734 for a related, but different issue

Note: the test case uses 0x0305 as the test version (to not require changes after TLS 1.3 goes final), behaviour can be changed by editing [line 605](
https://github.com/tomato42/tlsfuzzer/blob/7e58104dd1e4c36d45ced5e038bea60de10efff4/scripts/test-tls13-0rtt-garbage.py#L605)

   <details><summary>reproducer</summary><p>

```
openssl req -x509 -newkey rsa -keyout localhost.key -out localhost.crt -subj /CN=localhost -nodes -batch
openssl s_server -key localhost.key -cert localhost.crt -www 2>server.err >server.out &
openssl_pid=$!
git clone https://github.com/tomato42/tlsfuzzer
pushd tlsfuzzer
# won't be needed after https://github.com/tomato42/tlsfuzzer/pull/423 is merged
git checkout 0rtt-garbage-resumption
git clone https://github.com/tomato42/tlslite-ng .tlslite-ng
ln -s .tlslite-ng/tlslite tlslite
git clone https://github.com/warner/python-ecdsa .python-ecdsa
ln -s .python-ecdsa/ecdsa ecdsa
PYTHONPATH=. python scripts/test-tls13-0rtt-garbage.py
popd
kill $openssl_pid

```

  </p></details>

   <details><summary>tlsfuzzer output</summary><p>

note the "handshake with invalid 0-RTT and unknown version (downgrade to TLS 1.2)" test case
failure in "handshake with 0-RTT, HRR and early data after 2nd Client Hello" is caused by #6734 
```
sanity ...
OK

handshake with invalid 0-RTT and CCS ...
OK

handshake with invalid 0-RTT and HRR ...
OK

handshake with invalid 0-RTT with fragmented early data ...
OK

handshake with invalid 0-RTT and unknown version (downgrade to TLS 1.2) ...
Error encountered while processing node <tlsfuzzer.expect.ExpectChangeCipherSpec object at 0x7f20988a18d0> (child: <tlsfuzzer.expect.ExpectFinished object at 0x7f20988a1910>) with last message being: <tlslite.messages.Message object at 0x7f20988a6c10>
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-0rtt-garbage.py", line 727, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 210, in run
    RecordHeader2)))
AssertionError: Unexpected message from peer: Alert(fatal, unexpected_message)

undecryptable record later in handshake together with early_data ...
OK

handshake with invalid 0-RTT and CCS between early data records ...
OK

handshake with invalid PSK ...
OK

handshake with 0-RTT, HRR and early data after 2nd Client Hello ...
Error encountered while processing node <tlsfuzzer.expect.ExpectNewSessionTicket object at 0x7f209888a750> (child: <tlsfuzzer.expect.ExpectNewSessionTicket object at 0x7f209888a750>) with last message being: None
Error while processing
Traceback (most recent call last):
  File "scripts/test-tls13-0rtt-garbage.py", line 727, in main
    runner.run()
  File "/home/hkario/dev/tlsfuzzer/tlsfuzzer/runner.py", line 196, in run
    raise AssertionError("Unexpected closure from peer")
AssertionError: Unexpected closure from peer

handshake with invalid 0-RTT ...
OK

sanity ...
OK

Basic check if TLS 1.3 server can handle 0-RTT handshake
Verify that the server can handle a 0-RTT handshake from client
even if (or rather, especially if) it doesn't support 0-RTT.

version: 1

Test end
successful: 9
failed: 2
  'handshake with 0-RTT, HRR and early data after 2nd Client Hello'
  'handshake with invalid 0-RTT and unknown version (downgrade to TLS 1.2)'
```

   </p></details>

  <details><summary>openssl output</summary><p>

```
Using default temp DH parameters
ACCEPT
140464152045376:error:140940F5:SSL routines:ssl3_read_bytes:unexpected record:ssl/record/rec_layer_s3.c:1724:
140464152045376:error:1408F119:SSL routines:ssl3_get_record:decryption failed or bad record mac:ssl/record/ssl3_record.c:674:
```

  </p></details>

### PR DESCRIPTION
**Pull request by mattcaswell**
In some circumstances it is possible for a client to have a session
reporting a max early data value that is greater than the server will
support. In such cases the client could encounter an aborted connection.

Fixes #6735

### CODE REVIEW

File: doc/man3/SSL_read_early_data.pod [hunk 1]
```diff
@@ -267,6 +267,19 @@ Nagle's algorithm. If an application opts to disable Nagle's algorithm
...
 complete if appropriate.
 
+In rare circumstances, it may be possible for a client to have a session that
+reports a max early data value greater than 0, but where the server does not
+support this. For example, this can occur if a server has had its configuration
+changed to accept a lower max early data value such as by calling
+SSL_CTX_set_recv_max_early_data(). Another example is if a server used to
+support TLSv1.3 but was later downgraded to TLSv1.2. Sending early data to such
+a server will cause the connection to abort. Clients that encounter an aborted
+connection while sending early data may choose to retry the connection without
```
Comment (richsalz) [inline doc/man3/SSL_read_early_data.pod]: s/choose/want/

File: doc/man3/SSL_read_early_data.pod [hunk 2]
```diff
@@ -267,6 +267,19 @@ Nagle's algorithm. If an application opts to disable Nagle's algorithm
...
+sending early data. This does not happen automatically. A client will have to
```
Comment (richsalz) [inline doc/man3/SSL_read_early_data.pod]: s/. This/, as this/

File: doc/man3/SSL_read_early_data.pod [hunk 3]
```diff
@@ -267,6 +267,19 @@ Nagle's algorithm. If an application opts to disable Nagle's algorithm
...
+establish a new transport layer connection to the server and attempt the SSL/TLS
+connection again but without sending early data. Note that it is inadvisable to
```
Comment (richsalz) [inline doc/man3/SSL_read_early_data.pod]: s/is/might be/

Comment (mattcaswell) [reply doc/man3/SSL_read_early_data.pod]: Why only "might be"? Under what conditions would we advise it?

Comment (richsalz) [reply doc/man3/SSL_read_early_data.pod]: Suppose the server got a new revision that included TLS 1.3 support, but some of the new features caused it to crash.

Comment (tomato42) [reply doc/man3/SSL_read_early_data.pod]: that's a bug in that server and workarounding it should be applied either to none or all handshakes

**Comment by mattcaswell** [comment]
Updated commit pushed addressing the feedback so far. I didn't make the last change as I don't think we should be softening that statement. If that's a show stopper, an alternative would be to drop that sentence altogether.

**Comment by richsalz** [comment]
Not a show-stopper (I won't -1 it) but I won't approve and am fine if someone else does or if you remove the sentence.

**Comment by mattcaswell** [comment]
Ok - fair enough. Ideally I'd prefer to keep the sentence so I'll open it up to @openssl/committers, looking for another reviewer. If no one steps forward, I'll remove it.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

File: doc/man3/SSL_read_early_data.pod (hunk 3 above)
Comment (kaduk) [reply doc/man3/SSL_read_early_data.pod]: I'm not sure I correctly understand the senes of Rich's discomfort with this text.  FWIW, I would be happy to replace "inadvisable" with "forbidden by the spec".

**Pull request closed without merging**